Refreshable MV: DatabaseReplicated support, coordination among replicas

al13n321 opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Refreshable materialized views are now supported in Replicated databases.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

al13n321 committed:
Support cancelling DDL queries in DatabaseReplicated

al13n321 committed:
StopToken

al13n321 committed:
Refreshable MV: coordination and Replicated DB support

al13n321 committed:
Style

al13n321 committed:
Fix conflicts

al13n321 committed:
Conflict

fm4v commented:
RMV trigger with `APPEND` throws exception in `system.view_refreshes`: 
```
CREATE MATERIALIZED VIEW test_rmv
REFRESH EVERY 1 HOUR APPEND
ENGINE = Memory
AS SELECT
    now() AS a,
    number AS b
FROM numbers(2);

SELECT *
FROM system.view_refreshes;


Row 1:
──────
database:                 default
view:                     test_rmv
status:                   Scheduled
last_success_time:        ᴺᵁᴸᴸ
last_success_duration_ms: ᴺᵁᴸᴸ
last_refresh_time:        2024-04-29 11:57:21
last_refresh_replica:     
next_refresh_time:        2024-04-29 12:00:00
exception:                Code: 393. DB::Exception: There is no query or query context has expired. (THERE_IS_NO_QUERY), Stack trace (when copying this message, always include the lines below):

0. ./build_docker/./src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c99f6db
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000078069ac
2. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000007814d8b
3. ./build_docker/./src/Interpreters/Context.cpp:2396: DB::Context::getQueryContext() const @ 0x0000000010321089
4. ./build_docker/./src/Analyzer/Passes/QueryAnalysisPass.cpp:7414: DB::(anonymous namespace)::QueryAnalyzer::resolveTableFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&, DB::(anonymous namespace)::QueryExpressionsAliasVisitor&, bool) @ 0x0000000010a7434a
5. ./build_docker/./src/Analyzer/Passes/QueryAnalysisPass.cpp:7772: DB::(anonymous namespace)::QueryAnalyzer::resolveQueryJoinTreeNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::(anonymous namespace)::IdentifierResolveScope&, DB::(anonymous namespace)::QueryExpressionsAliasVisitor&) @ 0x0000000010a76d1c
6. ./build_docker/./src/Analyzer/Passes/QueryAnalysisPass.cpp:8021: DB::(anonymous namespace)::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x0000000010a64e91
7. ./build_docker/./src/Analyzer/Passes/QueryAnalysisPass.cpp:0: DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x0000000010a629e5
8. ./build_docker/./src/Analyzer/QueryTreePassManager.cpp:183: DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x0000000010a612a3
9. ./build_docker/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:0: DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) (.llvm.11603548317326480995) @ 0x0000000010cf40bd
10. ./build_docker/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:132: DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&) @ 0x0000000010cf2e51
11. ./build_docker/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:170: DB::InterpreterSelectQueryAnalyzer::getSampleBlock(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&) @ 0x0000000010cf4bfd
12. ./build_docker/./src/Storages/StorageMaterializedView.cpp:0: DB::StorageMaterializedView::prepareRefresh(bool, std::shared_ptr<DB::Context>, std::optional<DB::StorageID>&) const @ 0x0000000011661e09
13. ./build_docker/./src/Storages/MaterializedView/RefreshTask.cpp:528: DB::RefreshTask::refreshTask() @ 0x0000000012158fa2
14. ./base/glibc-compatibility/musl/clock_gettime.c:62: DB::BackgroundSchedulePool::threadFunction() @ 0x000000000fbf2765
15. ./build_docker/./src/Core/BackgroundSchedulePool.cpp:0: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000fbf3803
16. ./base/base/../base/wide_integer_impl.h:817: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000ca5972d
17. ? @ 0x000075b44ee94ac3
18. ? @ 0x000075b44ef26850
 (version 24.4.1.1793)
retry:                    0
progress:                 nan
read_rows:                0
read_bytes:               0
total_rows:               0
total_bytes:              0
written_rows:             0
written_bytes:            0
result_rows:              0
result_bytes:             0

```


version: `24.4.1.1793`

fm4v commented:
I can't quite understand how two units in `EVERY` work. There is no such problem with `AFTER`.

More cases: https://docs.google.com/document/d/1ncSaWYNMzUKkqUWpIoYlsJVQrhVC6fbEPTQhmka8cPQ/edit?usp=sharing
```
Interval: EVERY 1 YEAR
next_refresh_time 2025-01-01 CORRECT
predicted 2025-01-01 00:00:00
----
Interval: EVERY 1 YEAR 0 MONTH
next_refresh_time 2025-01-01 CORRECT
predicted 2025-01-01 00:00:00
Correct: True
—-
Interval: EVERY 1 YEAR 1 MONTH
next_refresh_time 2025-04-01 WRONG
predicted 2025-02-01 00:00:00
Correct: False
----
Interval: EVERY 1 YEAR 2 MONTH
next_refresh_time 2024-11-01 WRONG
predicted 2025-03-01 00:00:00
Correct: False
----
Interval: EVERY 1 YEAR 3 MONTH
next_refresh_time 2025-01-01 WRONG
predicted 2025-04-01 00:00:00
```

fm4v commented:
Should the same units be accepted like "EVERY 1 YEAR 2 YEAR"? Now they just being added together.

```
CREATE MATERIALIZED VIEW
test_rmv_schedule
REFRESH EVERY 1 YEAR 2 YEAR
ENGINE = Memory
EMPTY
AS SELECT now() as a, number as b FROM numbers(2)

----
current_time: 2024-04-30 15:24:14
Interval: EVERY 1 YEAR 2 YEAR
next_refresh_time 2027-01-01 00:00:00
```

```
CREATE MATERIALIZED VIEW
test_rmv_schedule
REFRESH AFTER 1 YEAR 2 YEAR
ENGINE = Memory
EMPTY
AS SELECT now() as a, number as b FROM numbers(2)

----
current_time: 2024-04-30 15:24:15
Interval: AFTER 1 YEAR 2 YEAR
next_refresh_time 2027-04-30 15:24:15
```

robot-clickhouse committed:
Automatic style fix

al13n321 committed:
Fix conflicts, 'context has expired', lint

al13n321 committed:
Reject repeated time units, add examples in documentation for EVERY

al13n321 commented:
> RMV trigger with APPEND throws exception in system.view_refreshes.

Fixed, thanks for catching it!

> Should the same units be accepted like "EVERY 1 YEAR 2 YEAR"? Now they just being added together.

Yeah, it wasn't clear what the user would expect, so I added them together because it's easy and seems natural (since different units are added together). Your comment made it more clear what the user expects  :) , so added a check to reject it.

---

> I can't quite understand how two and more units in EVERY work.

Maybe it was confusing because you tested it on Monday :) . `EVERY 1 WEEK` means refresh every Monday, even if the materialized view was created on a Wednesday. So the first refresh can be anywhere between 0 and 1 weeks from the view creation, depending on when it was created. Similarly, `EVERY 2 WEEKS` means Mondays of even-numbered weeks (counting from 1969-12-29), and the first refresh can be anywhere between 0 and 2 weeks from view creation.

This is the only reasonable-seeming way I could think of to generalize the notion of "start of each hour"/"every Monday"/"1st of every month" etc to ~arbitrary time intervals. Lmk if you have possibly-better ideas.

> I can't quite understand how two and more units in EVERY work.

They're converted to seconds or months and added together. Refresh happens when current time is divisible by that. Added 2 examples in `docs/en/sql-reference/statements/create/view.md`.

---

(Unnecessarily long explanation below, feel free to skip.)

The desired behavior for `EVERY 1 WEEK` is "every Monday". This has the properties: (a) it doesn't depend on when the materialized view was created, (b) consecutive refreshes are exactly 1 week apart. And non-properties: (c) the first refresh doesn't happen exactly 1 week after view creation, it can be anywhere between 0 and 1 week, (d) the schedule is not aligned with months or years.

I would expect `EVERY 2 WEEK` to have the same two properties and two non-properties, and also to fall on Mondays. There are exactly two possible ways to achieve that: all even-numbered Mondays or all odd-numbered Mondays. I just picked one.

The generalization to SECOND/MINUTE/HOUR/DAY/WEEK (not MONTH/YEAR) is: convert to seconds (say, `n` seconds), do a refresh when the number of seconds since 1969-12-29 (which is a Monday) is divisible by `n`. So:
 * `EVERY 1 <anything>` works as expected.
 * `EVERY 2/3/4/5/6/10/12/15/20/30/60 SECOND/MINUTE` work as expected. They're divisors of 60, so it's automatically aligned with minutes or hours.
 * `EVERY 2/3/4/6/8/12 HOUR` works as I guess people would expect: refresh at the start of every even-numbered/divisible-by-3-numbered/etc hour in the day.
 * When the interval is not a divisor of some bigger unit, refreshes won't be aligned with that bigger unit. Alternatively, we could forcefully align them to the next-bigger unit, making refresh interval non-uniform; e.g. EVERY 5 HOUR would refresh at 0, 5, 10, 15, and 20 hours after the start of each day, with a 4-hour interval between 20:00 and 00:00 of next day. That's what the original refreshable MV implementation did, but I didn't like it and changed it, i.e. traded property (b) for property (d). Lmk if you think we should change it back (and what to do with `EVERY 59 MINUTES` then).
 * It doesn't matter which units were used to specify the interval: `EVERY 1 WEEK` is the same as `EVERY 7 DAY` or `EVERY 168 HOUR` etc.
 * When the interval is just weird, like `EVERY 1 DAY 1 SECOND`, I'm guessing the user probably doesn't have any specific expectation for how exactly it'll work, so they won't be very surprised?

For MONTH/YEAR: convert to `n` months, refresh when current month number (since 1970-01) is divisible by `n`. Mixing seconds-based with months-based intervals is not allowed (e.g. `1 YEAR 2 WEEKS` - `Exception: Interval shouldn't contain both calendar units and clock units (e.g. months and days)`). Refreshing on day >28 of the month (e.g. `OFFSET 30 DAY`) is not allowed because February.

If you have suggestions for how and which of those things to document, based on what you found confusing - lmk. I'm hoping that the two new examples in `view.md` are enough to prompt the reader to think about this the intended way - absolute timestamps being divisible by time intervals.

pufit commented on src/Storages/MaterializedView/RefreshTask.cpp:
```diff
@@ -150,240 +241,301 @@ void RefreshTask::cancel()
+
+            lock.lock();
+            /// Re-check last_attempt_succeeded in case another refresh EXCHANGEd the table but failed to write its uuid to keeper.
+            throw_if_error();
```
We should probably add some backoff here.

pufit commented on tests/queries/0_stateless/02345_implicit_transaction.sql:
```diff
@@ -3,7 +3,7 @@
 CREATE TABLE target  (n Int64) engine=MergeTree order by n;
 CREATE MATERIALIZED VIEW landing_to_target TO target AS
-    SELECT n + throwIf(n == 3333)
+    SELECT n + throwIf(n == 3333) AS n
```
This is also concerning. Why are nonrefreshable views affected by the column's aliasing changes?

al13n321 replied:
Added compatibility setting in https://github.com/ClickHouse/ClickHouse/pull/58934 .

al13n321 committed:
Backoff and const

al13n321 commented on src/Interpreters/InterpreterCreateQuery.cpp:
```diff
@@ -908,6 +910,99 @@ void validateVirtualColumns(const IStorage & storage)
+void InterpreterCreateQuery::validateMaterializedViewColumnsAndEngine(const ASTCreateQuery & create, const TableProperties & properties, const DatabasePtr & database)
+{
+    /// This is not strict validation, just catches common errors that would make the view not work.
+    /// It's possible to circumvent these checks by ALTERing the view or target table after creation.
```
We probably should, I just documented this along the way.

al13n321 committed:
Conflicts

al13n321 committed:
Conflicts

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Small fixes

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Fixes

al13n321 committed:
Fix DatabaseReplicated executing queries out of order

al13n321 committed:
Fix test

al13n321 committed:
Fix test

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Fix progress reporting

al13n321 committed:
Unchange error message

al13n321 commented:
Ok, all the things appear to be fixed, @pufit now would be a good time for a review!

al13n321 committed:
Fixes

al13n321 committed:
Fix EXCHANGE race and do SYNC REPLICA when reading directly from target table if needed

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Conflict

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Conflict

al13n321 committed:
Conflict

al13n321 committed:
oops wrong pr

al13n321 committed:
Fix settings not being applied for getSampleBlock

al13n321 committed:
Merge remote-tracking branch 'origin/master' into HEAD

robot-clickhouse committed:
Automatic style fix

al13n321 commented:
AST fuzzer - known problem: https://github.com/ClickHouse/ClickHouse/issues/56640

al13n321 committed:
Unflaky test

al13n321 committed:
Merge branch 'mvrere' of github.com:ClickHouse/ClickHouse into mvrere

al13n321 committed:
Fix test flakiness

al13n321 commented:
Flaky tests: https://github.com/ClickHouse/ClickHouse/issues/62741 , https://github.com/ClickHouse/ClickHouse/issues/70282

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

al13n321 commented:
Flaky tests: https://github.com/ClickHouse/ClickHouse/issues/70337 https://github.com/ClickHouse/ClickHouse/issues/70169 https://github.com/ClickHouse/ClickHouse/issues/70319

al13n321 committed:
Merge remote-tracking branch 'origin/master' into mvrere

clickgapai commented:
Hi @al13n321 @pufit — while reviewing this PR I found the following:

- **Coverage gap:** `TO INNER UUID` rejected for refreshable MV without APPEND — new BAD_ARGUMENTS check untested — https://github.com/ClickHouse/ClickHouse/pull/103894

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.
